diff --git a/piton/design/chip/tile/ariane b/piton/design/chip/tile/ariane
--- a/piton/design/chip/tile/ariane
+++ b/piton/design/chip/tile/ariane
@@ -1 +1 @@
-Subproject commit cbbcf3565c3829b40288848e2a6a49be689e7922
+Subproject commit cbbcf3565c3829b40288848e2a6a49be689e7922-dirty
diff --git a/piton/design/chipset/include/mc_define.h b/piton/design/chipset/include/mc_define.h
index 1338770..24371d2 100644
--- a/piton/design/chipset/include/mc_define.h
+++ b/piton/design/chipset/include/mc_define.h
@@ -67,6 +67,44 @@
     `define DDR3_CKE_WIDTH          1
     `define DDR3_CS_WIDTH           1
     `define DDR3_ODT_WIDTH          1
+`elsif DUALVU440_BOARD
+    `define BOARD_MEM_SIZE_MB       8192
+    `define WORDS_PER_BURST         8
+    `define WORD_SIZE               8 // in bytes
+    `define MIG_APP_ADDR_WIDTH      30
+    `define MIG_APP_CMD_WIDTH       3
+    `define MIG_APP_DATA_WIDTH      512
+    `define MIG_APP_MASK_WIDTH      64
+
+    `define DDR3_DQ_WIDTH           64
+    `define DDR3_DQS_WIDTH          8
+    `define DDR3_ADDR_WIDTH         17
+    `define DDR3_BA_WIDTH           2
+    `define DDR3_DM_WIDTH           8
+    `define DDR3_CK_WIDTH           1
+    `define DDR3_CKE_WIDTH          1
+    `define DDR3_CS_WIDTH           1
+    `define DDR3_BG_WIDTH           2
+    `define DDR3_ODT_WIDTH          1
+`elsif VU19P_BOARD
+    `define BOARD_MEM_SIZE_MB       8192
+    `define WORDS_PER_BURST         8
+    `define WORD_SIZE               8 // in bytes
+    `define MIG_APP_ADDR_WIDTH      30
+    `define MIG_APP_CMD_WIDTH       3
+    `define MIG_APP_DATA_WIDTH      512
+    `define MIG_APP_MASK_WIDTH      64
+
+    `define DDR3_DQ_WIDTH           64
+    `define DDR3_DQS_WIDTH          8
+    `define DDR3_ADDR_WIDTH         17
+    `define DDR3_BA_WIDTH           2
+    `define DDR3_DM_WIDTH           8
+    `define DDR3_CK_WIDTH           1
+    `define DDR3_CKE_WIDTH          1
+    `define DDR3_CS_WIDTH           1
+    `define DDR3_BG_WIDTH           2
+    `define DDR3_ODT_WIDTH          1    
 `elsif VCU118_BOARD
     `define BOARD_MEM_SIZE_MB       2048
     `define WORDS_PER_BURST         8
diff --git a/piton/design/chipset/noc_axi4_bridge/rtl/noc_axi4_bridge_ser.v b/piton/design/chipset/noc_axi4_bridge/rtl/noc_axi4_bridge_ser.v
index 6885310..9a05761 100644
--- a/piton/design/chipset/noc_axi4_bridge/rtl/noc_axi4_bridge_ser.v
+++ b/piton/design/chipset/noc_axi4_bridge/rtl/noc_axi4_bridge_ser.v
@@ -48,7 +48,7 @@ module noc_axi4_bridge_ser(
 localparam ACCEPT = 2'd0;
 localparam SEND_HEADER = 2'd1;
 localparam SEND_DATA = 2'd2;
-
+reg [`NOC_DATA_WIDTH-1:0] resp_header;
 reg [`AXI4_DATA_WIDTH-1:0] data_in_f;
 
 wire in_go = in_val & in_rdy;
@@ -118,7 +118,7 @@ always @(posedge clk) begin
   end
 end
 
-reg [`NOC_DATA_WIDTH-1:0] resp_header;
+
 always @(posedge clk) begin
   if (~rst_n) begin
     resp_header <= `NOC_DATA_WIDTH'b0;
diff --git a/piton/design/chipset/noc_sd_bridge/rtl/sd_cmd_serial_host.v b/piton/design/chipset/noc_sd_bridge/rtl/sd_cmd_serial_host.v
index 36c6b96..650fe2f 100644
--- a/piton/design/chipset/noc_sd_bridge/rtl/sd_cmd_serial_host.v
+++ b/piton/design/chipset/noc_sd_bridge/rtl/sd_cmd_serial_host.v
@@ -74,9 +74,18 @@ output reg [119:0] response_o;
 output reg finish_o;
 output reg crc_ok_o;
 output reg index_ok_o;
+
+
+`ifdef DUALVU440_BOARD
+output reg cmd_oe_o;
+output reg cmd_out_o;
+
+`else
 (* iob="true" *) output reg cmd_oe_o;
 (* iob="true" *) output reg cmd_out_o;
 
+`endif
+
 //-------------Internal Constant-------------
 parameter INIT_DELAY = 4;
 parameter BITS_TO_SEND = 48;
@@ -84,7 +93,12 @@ parameter CMD_SIZE = 40;
 parameter RESP_SIZE = 128;
 
 //---------------Internal variable-----------
+
+`ifdef DUALVU440_BOARD
+reg cmd_dat_reg;
+`else
 (* iob="true" *) reg cmd_dat_reg;
+`endif
 reg cmd_out_o_reg;
 reg cmd_oe_o_reg;
 integer resp_len;
diff --git a/piton/design/chipset/noc_sd_bridge/rtl/sd_data_serial_host.v b/piton/design/chipset/noc_sd_bridge/rtl/sd_data_serial_host.v
index e6b203d..964b546 100644
--- a/piton/design/chipset/noc_sd_bridge/rtl/sd_data_serial_host.v
+++ b/piton/design/chipset/noc_sd_bridge/rtl/sd_data_serial_host.v
@@ -57,9 +57,14 @@ module sd_data_serial_host(
            //Rx Fifo
            output reg [31:0] data_out,
            output reg we,
+           `ifdef DUALVU440_BOARD
            //tristate data
+           output reg DAT_oe_o,
+           output reg[3:0] DAT_dat_o,
+           `else
            (* iob="true" *) output reg DAT_oe_o,
            (* iob="true" *) output reg[3:0] DAT_dat_o,
+           `endif
            input [3:0] DAT_dat_i,
            //Controll signals
            input [`BLKSIZE_W-1:0] blksize,
@@ -71,8 +76,11 @@ module sd_data_serial_host(
            output busy,
            output reg crc_ok
        );
-
+`ifdef DUALVU440_BOARD
+reg [3:0] DAT_dat_reg;
+`else
 (* iob="true" *) reg [3:0] DAT_dat_reg;
+`endif
 reg [3:0] DAT_dat_o_reg;
 reg DAT_oe_o_reg;
 reg [`BLKSIZE_W-1+3:0] data_cycles;
diff --git a/piton/design/chipset/rtl/chipset.v b/piton/design/chipset/rtl/chipset.v
index d5e9148..3a38d88 100644
--- a/piton/design/chipset/rtl/chipset.v
+++ b/piton/design/chipset/rtl/chipset.v
@@ -455,6 +455,18 @@ module chipset(
         input                                               btnr,
         input                                               btnu,
         input                                               btnd,
+        input                                               btnc,  
+    `elsif DUALVU440_BOARD
+        input                                               btnl,
+        input                                               btnr,
+        input                                               btnu,
+        input                                               btnd,
+        input                                               btnc,  
+    `elsif VU19P_LS_BOARD
+        input                                               btnl,
+        input                                               btnr,
+        input                                               btnu,
+        input                                               btnd,
         input                                               btnc,           
     `endif
 
@@ -462,6 +474,12 @@ module chipset(
     `ifdef VCU118_BOARD
         // we only have 4 gpio dip switches on this board
         input  [3:0]                                        sw,
+    `elsif DUALVU440_BOARD
+        // we only have 4 gpio dip switches on this board
+        input  [3:0]                                        sw,
+    `elsif VU19P_LS_BOARD
+        // we only have 4 gpio dip switches on this board
+        input  [3:0]                                        sw,
     `elsif XUPP3R_BOARD
         // no switches :(
     `else         
@@ -748,6 +766,12 @@ end
             `ifdef VCU118_BOARD
                 assign uart_boot_en    = sw[0];
                 assign uart_timeout_en = sw[1];
+            `elsif DUALVU440_BOARD
+                assign uart_boot_en    = sw[0];
+                assign uart_timeout_en = sw[1];
+            `elsif VU19P_LS_BOARD
+                assign uart_boot_en    = sw[0];
+                assign uart_timeout_en = sw[1];
             `elsif XUPP3R_BOARD
                 assign uart_boot_en    = 1'b1;
                 assign uart_timeout_en = 1'b0;
@@ -763,6 +787,12 @@ end
     `ifdef VCU118_BOARD
         // only two switches available...
         assign noc_power_test_hop_count = {2'b0, sw[3:2]};
+    `elsif DUALVU440_BOARD
+        // only two switches available...
+        assign noc_power_test_hop_count = {4'b0};
+    `elsif VU19P_LS_BOARD
+        // only two switches available...
+        assign noc_power_test_hop_count = {4'b0};
     `elsif XUPP3R_BOARD
         // no switches :(
         assign noc_power_test_hop_count = 4'b0;
@@ -1574,6 +1604,22 @@ chipset_impl_noc_power_test  chipset_impl (
             .D2(0),
             .SR(0)
             );
+    `elsif DUALVU440_BOARD
+        ODDRE1 sd_clk_oddr (
+            .Q(sd_clk_out),
+            .C(sd_clk_out_internal),
+            .D1(1),
+            .D2(0),
+            .SR(0)
+            );
+    `elsif VU19P_LS_BOARD
+        ODDRE1 sd_clk_oddr (
+            .Q(sd_clk_out),
+            .C(sd_clk_out_internal),
+            .D1(1),
+            .D2(0),
+            .SR(0)
+            );
     `else 
         ODDR sd_clk_oddr (
             .Q(sd_clk_out),
diff --git a/piton/design/chipset/xilinx/vcu118/devices_ariane.xml b/piton/design/chipset/xilinx/vcu118/devices_ariane.xml
index 9f2e1c3..d4eb3f5 100644
--- a/piton/design/chipset/xilinx/vcu118/devices_ariane.xml
+++ b/piton/design/chipset/xilinx/vcu118/devices_ariane.xml
@@ -22,8 +22,8 @@ Description: Peripheral address map for OpenPiton+Ariane configurations.
     <port>
         <name>mem</name>
         <base>0x80000000</base>
-        <!-- 4GB -->
-        <length>0x100000000</length>
+        <!-- 8GB -->                
+        <length>0x200000000</length>
     </port>
     <port>
         <name>iob</name>
diff --git a/piton/design/include/piton_system.vh b/piton/design/include/piton_system.vh
index 5b6618f..e980850 100644
--- a/piton/design/include/piton_system.vh
+++ b/piton/design/include/piton_system.vh
@@ -118,6 +118,10 @@
     `define PITON_CHIPSET_DIFF_CLK
 `elsif VCU118_BOARD
     `define PITON_CHIPSET_DIFF_CLK
+`elsif DUALVU440_BOARD
+    `define PITON_CHIPSET_DIFF_CLK
+`elsif VU19P_LS_BOARD
+    `define PITON_CHIPSET_DIFF_CLK
 `elsif XUPP3R_BOARD
     `define PITON_CHIPSET_DIFF_CLK
 `elsif GENESYS2_BOARD
@@ -157,6 +161,10 @@
 
 `ifdef VCU118_BOARD
     `define PITONSYS_DDR4
+`elsif DUALVU440_BOARD
+    `define PITONSYS_DDR4
+`elsif VU19P_LS_BOARD
+    `define PITONSYS_DDR4
 `elsif XUPP3R_BOARD
     `define PITONSYS_DDR4
 `endif
diff --git a/piton/design/rtl/system.v b/piton/design/rtl/system.v
index 004cb11..5228a40 100644
--- a/piton/design/rtl/system.v
+++ b/piton/design/rtl/system.v
@@ -178,6 +178,8 @@ module system(
 `ifdef PITON_ARIANE
 `ifndef VC707_BOARD
 `ifndef VCU118_BOARD
+`ifndef DUALVU440_BOARD
+`ifndef VU19P_LS_BOARD
 `ifndef NEXYSVIDEO_BOARD
 `ifndef XUPP3R_BOARD
 `ifndef F1_BOARD
@@ -190,6 +192,8 @@ module system(
 `endif//XUPP3R_BOARD
 `endif //NEXYSVIDEO_BOARD
 `endif //VCU118_BOARD
+`endif //VCU118_BOARD
+`endif //VCU118_BOARD
 `endif  //VC707_BOARD
 `endif //PITON_ARIANE
 `endif //PITON_FPGA_SYNTH
@@ -309,14 +313,29 @@ module system(
 		input                                       uart_cts,
 		output                                      uart_rts,
 `endif // VCU118_BOARD
+
+`ifdef DUALVU440_BOARD
+        input                                       uart_cts,
+        output                                      uart_rts,
+`endif // DUALVU440_BOARD
+
+`ifdef VU19P_LS_BOARD
+        input                                       uart_cts,
+        output                                      uart_rts,
+`endif // VU19P_LS_BOARD
+
 `endif // endif PITONSYS_UART
 
 `ifdef PITONSYS_SPI
     `ifndef VC707_BOARD
     input                                       sd_cd,
-    `ifndef VCU118_BOARD
+        `ifndef VCU118_BOARD
+        `ifndef DUALVU440_BOARD
+        `ifndef VU19_LS_BOARD
     output                                      sd_reset,
-    `endif
+        `endif
+        `endif    
+        `endif
     `endif
     output                                      sd_clk_out,
     inout                                       sd_cmd,
@@ -379,10 +398,26 @@ module system(
     input                                       btnu,
     input                                       btnd,
     input                                       btnc,
+`elsif DUAL440_BOARD
+    input                                       btnl,
+    input                                       btnr,
+    input                                       btnu,
+    input                                       btnd,
+    input                                       btnc,
+`elsif VU19P_LS_BOARD
+    input                                       btnl,
+    input                                       btnr,
+    input                                       btnu,
+    input                                       btnd,
+    input                                       btnc,
 `endif
 
 `ifdef VCU118_BOARD
     // we only have 4 gpio dip switches on this board
+    input  [3:0]                                sw,    
+`elsif DUALVU440_BOARD
+    input  [3:0]                                sw,
+`elsif VU19P_LS_BOARD
     input  [3:0]                                sw,
 `elsif XUPP3R_BOARD
     // no switches :(
@@ -553,6 +588,17 @@ assign rtc = rtc_div[6];
 assign uart_rts = 1'b0;
 `endif // VCU118_BOARD
 
+`ifdef DUALVU440_BOARD
+// tie off
+assign uart_rts = 1'b0;
+`endif // VCU118_BOARD
+
+`ifdef VU19P_LS_BOARD
+// tie off
+assign uart_rts = 1'b0;
+`endif // VCU118_BOARD
+
+
 // Different reset active levels for different boards
 always @ *
 begin
@@ -1135,10 +1181,15 @@ chipset chipset(
 `ifdef PITONSYS_SPI
     `ifndef VC707_BOARD
     .sd_cd(sd_cd),
-    `ifndef VCU118_BOARD
+    `ifndef VCU118_BOARD    
+    `ifndef DUALVU440_BOARD
+    `ifndef VU19P_LS_BOARD
     .sd_reset(sd_reset),
     `endif
     `endif
+    `endif
+
+    `endif
     .sd_clk_out(sd_clk_out),
     .sd_cmd(sd_cmd),
     .sd_dat(sd_dat),
@@ -1187,6 +1238,18 @@ chipset chipset(
     .btnu(btnu),
     .btnd(btnd),
     .btnc(btnc),
+`elsif DUALVU440_BOARD
+    .btnl(btnl),
+    .btnr(btnr),
+    .btnu(btnu),
+    .btnd(btnd),
+    .btnc(btnc),
+`elsif VU19P_LS_BOARD
+    .btnl(btnl),
+    .btnr(btnr),
+    .btnu(btnu),
+    .btnd(btnd),
+    .btnc(btnc),
 `endif
 
 `ifndef XUPP3R_BOARD
diff --git a/piton/piton_settings.bash b/piton/piton_settings.bash
index 79754ad..37036e3 100644
--- a/piton/piton_settings.bash
+++ b/piton/piton_settings.bash
@@ -144,3 +144,4 @@ export M4PATH=$DV_ROOT/tools/$OS/$CPU/lib/m4
 
 # Set library path for the new goldfinger
 export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$DV_ROOT/tools/src/goldfinger/lib
+function vcs() { command vcs -full64 "$@"; }; export -f vcs
\ No newline at end of file
diff --git a/piton/tools/src/proto/block.list b/piton/tools/src/proto/block.list
index 04a3323..8a9459f 100644
--- a/piton/tools/src/proto/block.list
+++ b/piton/tools/src/proto/block.list
@@ -25,8 +25,8 @@
 # Format:
 # BlockID                   BlockPath                           Supported Board,Frequency(MHz),DDRSize(Mbytes)
 piton_aws                   ../../build/f1/piton_aws/design                         f1,62.5,4096
-system                      .                                                       vc707,60,1024;genesys2,66.667,1024;nexysVideo,30,512;vcu118,100,2048;xupp3r,60,32768
-chipset                     chipset                                                 genesys2,66.667,1024;piton_board,50,0
+system                      .                                                       vc707,60,1024;genesys2,66.667,1024;nexysVideo,30,512;vcu118,100,2048;dualvu440,50,8192;vu19p_ls,50,8192;xupp3r,60,32768
+chipset                     chipset                                                 genesys2,66.667,1024;piton_board,50
 passthru                    passthru                                                piton_board,100,0
 passthru_loopback           fpga_tests/passthru_loopback                            piton_board,100,0
 chip                        chip                                                    genesys2,66.667,1024
diff --git a/piton/tools/src/proto/board.list b/piton/tools/src/proto/board.list
index 6149781..9e2132c 100644
--- a/piton/tools/src/proto/board.list
+++ b/piton/tools/src/proto/board.list
@@ -29,4 +29,6 @@ vc707               vivado
 nexysVideo          vivado
 f1                  vivado
 vcu118              vivado
+dualvu440           vivado
+vu19p_ls			vivado
 xupp3r              vivado
diff --git a/piton/tools/src/proto/common/setup.tcl b/piton/tools/src/proto/common/setup.tcl
index bc08ae3..aa4d765 100644
--- a/piton/tools/src/proto/common/setup.tcl
+++ b/piton/tools/src/proto/common/setup.tcl
@@ -110,13 +110,19 @@ for {set k 0} {$k < $::env(PITON_NUM_TILES)} {incr k} {
   }
 }
 
-puts "INFO: Using Defines: ${ALL_DEFAULT_VERILOG_MACROS}"
+set tmp_PYTHONPATH $env(PYTHONPATH)
+set tmp_PYTHONHOME $env(PYTHONHOME)
 
+puts "INFO: Using Defines: ${ALL_DEFAULT_VERILOG_MACROS}"
+  unset ::env(PYTHONPATH)    
+  unset ::env(PYTHONHOME)
 # Pre-process PyHP files
 source $DV_ROOT/tools/src/proto/common/pyhp_preprocess.tcl
 set ALL_RTL_IMPL_FILES [pyhp_preprocess ${ALL_RTL_IMPL_FILES}]
 set ALL_INCLUDE_FILES [pyhp_preprocess ${ALL_INCLUDE_FILES}]
 
+ set ::env(PYTHONPATH) $tmp_PYTHONPATH
+ set ::env(PYTHONHOME) $tmp_PYTHONHOME
 
 if  {[info exists ::env(PITON_ARIANE)]} {
   puts "INFO: compiling DTS and bootroms for Ariane (MAX_HARTS=$::env(PITON_NUM_TILES), UART_FREQ=$env(CONFIG_SYS_FREQ))..."
@@ -126,17 +132,25 @@ if  {[info exists ::env(PITON_ARIANE)]} {
   # otherwise this command fails...
   exec make clean 2> /dev/null
   exec make all 2> /dev/null
+
+#  if { [  catch {exec make all 2> /dev/null}  msg] } {
+#   puts "Something seems to have gone wrong:"
+#   puts "Information about it: $::errorInfo"  }
+
+  #exec make all 
   cd $::env(ARIANE_ROOT)/openpiton/bootrom/linux
   # Note: dd dumps info to stderr that we do not want to interpret
   # otherwise this command fails...
   exec make clean 2> /dev/null
-  exec make all MAX_HARTS=$::env(PITON_NUM_TILES) UART_FREQ=$::env(CONFIG_SYS_FREQ) 2> /dev/null
+  catch {exec make all MAX_HARTS=$::env(PITON_NUM_TILES) UART_FREQ=$::env(CONFIG_SYS_FREQ) 2> /dev/null}
   puts "INFO: done"
   # two targets per hart (M,S) and two interrupt sources (UART, Ethernet)
   set NUM_TARGETS [expr 2*$::env(PITON_NUM_TILES)]
   set NUM_SOURCES 2
   puts "INFO: generating PLIC for Ariane ($NUM_TARGETS targets, $NUM_SOURCES sources)..."
-  cd $::env(ARIANE_ROOT)/src/rv_plic/rtl
+  unset ::env(PYTHONPATH)    
+  unset ::env(PYTHONHOME)
+  cd $::env(ARIANE_ROOT)/src/rv_plic/rtl  
   exec ./gen_plic_addrmap.py -t $NUM_TARGETS -s $NUM_SOURCES > plic_regmap.sv
 
   cd $TMP
diff --git a/piton/tools/src/proto/fpga_lib.py b/piton/tools/src/proto/fpga_lib.py
index f6875b1..6b5182f 100644
--- a/piton/tools/src/proto/fpga_lib.py
+++ b/piton/tools/src/proto/fpga_lib.py
@@ -48,6 +48,8 @@ MAP_MODULE_NAME = "storage_addr_trans.v"
 NOC_PAYLOAD_WIDTH = 512
 STORAGE_BLOCK_BIT_WIDTH         =   {   "ddr":  {   "vc707":512,
                                                     "vcu118":512,
+                                                    "dualvu440":512,
+                                                    "vu19p_ls":512,
                                                     "xupp3r":512,
                                                     "nexys4ddr":128,
                                                     "genesys2":256,
@@ -56,6 +58,8 @@ STORAGE_BLOCK_BIT_WIDTH         =   {   "ddr":  {   "vc707":512,
                                                 },
                                         "bram": {   "vc707":512,
                                                     "vcu118":512,
+                                                    "dualvu440":512,
+                                                    "vu19p_ls":512,
                                                     "xupp3r":512,
                                                     "nexys4ddr":512,
                                                     "genesys2":512,
@@ -65,6 +69,8 @@ STORAGE_BLOCK_BIT_WIDTH         =   {   "ddr":  {   "vc707":512,
                                                 },
                                         "dmw":  {   "vc707":512,
                                                     "vcu118":512,
+                                                    "dualvu440":512,
+                                                    "vu19p_ls":512,
                                                     "xupp3r":512,
                                                     "nexys4ddr":512,
                                                     "genesys2":512,
@@ -76,6 +82,8 @@ STORAGE_BLOCK_BIT_WIDTH         =   {   "ddr":  {   "vc707":512,
 
 STORAGE_ADDRESSABLE_BIT_WIDTH   =   {   "ddr":  {   "vc707":64,
                                                     "vcu118":64,
+                                                    "dualvu440":64,
+                                                    "vu19p_ls":64,
                                                     "xupp3r":64,
                                                     "nexys4ddr":16,
                                                     "genesys2":32,
@@ -84,6 +92,8 @@ STORAGE_ADDRESSABLE_BIT_WIDTH   =   {   "ddr":  {   "vc707":64,
                                                 },
                                         "bram": {   "vc707":512,
                                                     "vcu118":512,
+                                                    "dualvu440":512,
+                                                    "vu19p_ls":512,
                                                     "xupp3r":512,
                                                     "nexys4ddr":512,
                                                     "genesys2":512,
@@ -93,6 +103,8 @@ STORAGE_ADDRESSABLE_BIT_WIDTH   =   {   "ddr":  {   "vc707":64,
                                                 },
                                         "dmw": {    "vc707":512,
                                                     "vcu118":512,
+                                                    "dualvu440":512,
+                                                    "vu19p_ls":512,
                                                     "xupp3r":512,
                                                     "nexys4ddr":512,
                                                     "genesys2":512,
@@ -104,6 +116,8 @@ STORAGE_ADDRESSABLE_BIT_WIDTH   =   {   "ddr":  {   "vc707":64,
 
 STORAGE_BIT_SIZE                =   {   "ddr":  {   "vc707":8*2**30,
                                                     "vcu118":2*8*2**30,
+                                                    "dualvu440":8*8*2**30,
+                                                    "vu19p_ls":8*8*2**30,
                                                     "xupp3r":32*8*2**30,
                                                     "nexys4ddr":8*128*2**20,
                                                     "genesys2":8*2**30,
@@ -112,6 +126,8 @@ STORAGE_BIT_SIZE                =   {   "ddr":  {   "vc707":8*2**30,
                                                 },
                                         "bram": {   "vc707":16384*512,
                                                     "vcu118":16384*512,
+                                                    "dualvu440":16384*512,
+                                                    "vu19p_ls":16384*512,
                                                     "xupp3r":16384*512,
                                                     "nexys4ddr":16384*512,
                                                     "genesys2":16384*512,
@@ -121,6 +137,8 @@ STORAGE_BIT_SIZE                =   {   "ddr":  {   "vc707":8*2**30,
                                                 },
                                         "dmw":  {   "vc707":8*2**30,
                                                     "vcu118":2*8*2**30,
+                                                    "dualvu440":8*8*2**30,
+                                                    "vu19p_ls":8*8*2**30,
                                                     "xupp3r":32*8*2**30,
                                                     "nexys4ddr":8*128*2**20,
                                                     "genesys2":8*2**30,
diff --git a/piton/tools/src/proto/image2stream.py b/piton/tools/src/proto/image2stream.py
index 2e8f071..4c72743 100644
--- a/piton/tools/src/proto/image2stream.py
+++ b/piton/tools/src/proto/image2stream.py
@@ -1,3 +1,4 @@
+  
 # Copyright (c) 2017 Princeton University
 # All rights reserved.
 # 
@@ -55,7 +56,7 @@ def flushGroup(fptr, addr_str, gr_blocks):
 def strFromAddr(addr, width):
     s = str(hex(addr))
     h = s[2:]
-    return (width-len(h))*'0' + h
+    return (int(width)-len(h))*'0' + h
 
 def makeStreamFile(fname_bram):
     fin = open(fname_bram, 'r')
@@ -112,4 +113,5 @@ def makeStreamFile(fname_bram):
     # print "DEBUG: # groups with overflow: %d" % block_ovrfl
 
     fin.close()
-    fout.close()
\ No newline at end of file
+    fout.close()
+
diff --git a/piton/tools/src/proto/pitonstream,1.0 b/piton/tools/src/proto/pitonstream,1.0
index a89bea5..3acae1c 100755
--- a/piton/tools/src/proto/pitonstream,1.0
+++ b/piton/tools/src/proto/pitonstream,1.0
@@ -37,7 +37,7 @@
 #
 #####################################################################
 
-import time, re, sys, os, serial, shlex, subprocess, csv
+import time, re, sys, os, serial, shlex, subprocess, csv, binascii
 import make_mem_map as mmm
 import image2stream as i2s
 from optparse import OptionParser
@@ -93,6 +93,7 @@ def usage():
     print("              Name of a supported core configuration:", file=sys.stderr)
     print("                  sparc     (opensparc T1 core, default)", file=sys.stderr)
     print("                  ariane    (64bit RISCV core)", file=sys.stderr)
+    print("                  blackparrot (64bit RISCV core)", file=sys.stderr)
     print("\n       -f, --file <filename>", file=sys.stderr)
     print("              File name with test names. If option --ustr is not specified,", file=sys.stderr)
     print("              they should be assembly test names, otherwise list of generated", file=sys.stderr)
@@ -191,7 +192,7 @@ def pollUART(ser):
         if len(rdata) > 0:
             # have to store all data in case of a key
             # is split between two different reads from serial device
-            rdata_all += rdata
+            rdata_all += str(rdata)
 
             was_key = False
             for key in list(KEY_EVENT_MAP.keys()):
@@ -202,7 +203,7 @@ def pollUART(ser):
                     event = KEY_EVENT_MAP[key]
                     # if keyword was in the current read chunk of data,
                     # we can remove it, since it's not test output
-                    rdata = re.sub("%s" % key, "", rdata)
+                    rdata = re.sub("%s" % key, "", str(rdata))
                     was_key = True
                     break
 
@@ -230,7 +231,7 @@ def loadTest(tname, ser):
     prev_perc = 0
     for line in f:
         line_striped = line.strip()
-        ser.write(line_striped.decode("hex"))
+        ser.write(binascii.unhexlify(line_striped))
         line_cnt += 1
         perc = 100*line_cnt/num_lines
         if (perc > prev_perc):
@@ -304,7 +305,7 @@ def compileASMTest(tname, uart_div_latch, st_brd, flog, midas_args=None, coreTyp
 
     # Since we are using an existent synthesized mapping, we have to check
     # that all sections of new tests are mapped
-    rv = isTranslatorOK(addr_data_map, flog, coreType=="ariane")
+    rv = isTranslatorOK(addr_data_map, flog, coreType in ["ariane", "blackparrot"])
     if not rv:
         print_error("Can't use existent storage_addr_trans.tmp.v for %s" % tname)
         return 1
@@ -379,20 +380,20 @@ def main():
 
     # Wait until FPGA is configure
     if not options.no_wait_fpga_config:
-    	writeSTDOUT("Press reset button on FPGA\n")
-    	writeSTDOUT("Waiting...\n")
-    	while True:
-    	    try:
-    	        event = pollUART(ser)[0]
-    	        if event == CFG_DONE_EVENT:
-    	            writeSTDOUT("\nConfiguration is complete\n\n")
-    	            break
-    	    except KeyboardInterrupt:
-    	        processKeyboardInterrupt(ser, flog)
-    	        return
+        writeSTDOUT("Press reset button on FPGA\n")
+        writeSTDOUT("Waiting...\n")
+        while True:
+            try:
+                event = pollUART(ser)[0]
+                if event == CFG_DONE_EVENT:
+                    writeSTDOUT("\nConfiguration is complete\n\n")
+                    break
+            except KeyboardInterrupt:
+                processKeyboardInterrupt(ser, flog)
+                return
 
     # Main test loop
-    fp = open("pitonstream_summary.csv", "wb")
+    fp = open("pitonstream_summary.csv", "w")
     csvwriter = csv.writer(fp, delimiter=',', quotechar='|')
     csvwriter.writerow(["seqID","test","status","walltime (seconds)"])
     test_num = len(tests)
@@ -400,8 +401,10 @@ def main():
     for tname in tests:
         test_cnt += 1
         print_info("Running %s: %d out of %d test" % (tname, test_cnt, test_num))
-        if options.precompiled and options.core == "Ariane":
+        if options.precompiled and options.core == "ariane":
             print_info("Using precompiled RISCV test for Ariane.")
+        if options.precompiled and options.core == "blackparrot":
+            print_info("Using precompiled RISCV test for BlackParrot.")
         # Compile .s tests to get .ustr
         if options.ustr == False:
             rv = compileASMTest(tname, uart_div_latch, st_brd, flog, options.midas_args, options.core, options.precompiled, options.x_tiles, options.y_tiles)
@@ -457,4 +460,4 @@ def main():
     exitProgram(0, ser, flog)
 
 if __name__ == '__main__':
-    main()
+    main()
\ No newline at end of file
diff --git a/piton/tools/src/proto/protosyn,2.5 b/piton/tools/src/proto/protosyn,2.5
index 0771522..a4d7785 100755
--- a/piton/tools/src/proto/protosyn,2.5
+++ b/piton/tools/src/proto/protosyn,2.5
@@ -60,6 +60,8 @@ def usage():
     print("              Name of a supported Xilinx's development board. Available options are:", file=sys.stderr)
     print("                  vc707", file=sys.stderr)
     print("                  vcu118", file=sys.stderr)
+    print("                  dualvu440", file=sys.stderr)
+    print("                  vu19p_ls", file=sys.stderr)    
     print("                  xupp3r", file=sys.stderr)
     print("                  genesys2", file=sys.stderr)
     print("                  nexysVideo", file=sys.stderr)
